types-core/index.d.ts

```diff
@@ -18,11 +18,13 @@ declare module '@azure/functions-core' {
      */
     export function registerHook(hookName: 'preInvocation', callback: PreInvocationCallback): Disposable;
     export function registerHook(hookName: 'postInvocation', callback: PostInvocationCallback): Disposable;
+    export function registerHook(hookName: 'appStartup', callback: AppStartupCallback): Disposable;
```

Comment:
I look at fastify, and it looks like their hooks generally follow the `onEvent` pattern, which we don't do for the invocation hooks. It looks like their most equivalent hook is called `onReady`. I asked the Python team and you're right that they don't have this hook. They have an app-level level, but that is essentially an invocation hook that is by default run for _all_ functions, as opposed to their function-level hooks, which are specific to a particular function.

Other potential nouns we could use instead of "startup":
- start
- load
- init/initialize
- ready

Could definitely use suggestions from you or @lilyjma though.

I don't have a strong preference for keeping the "app" prefix by name, but I would like to have some way of telling that this hook will be run only once the function _app_ starts, not the function (i.e. a pre-invocation hook).

Reply:
I vote for `appStart` or `appInit`. Some reasoning:

### don't like
- load: This isn't broad enough - to me we also "execute" code and "load" doesn't include that
- initialize: word just seems too long
- ready: It just doesn't feel clear to me when an app is considered "ready"
- startup: I don't like words where you're not sure if it's one word or two words (startup vs startUp vs start-up) and IMO english has too many nuances to the words "up" vs "down" which make it confusing for non-native speakers

### do like
- app prefix: I agree it's nice to have a way to clarify this is for the app not the function
- init: This matches the rpc terminology (WorkerInitRequest) and feels accurate. Only downside is I'm not a huge fan of abbreviations by default, but I think most people know "init"
- start: feels simple, short, and obvious in the meaning. Downside is I don't have an example to point to saying "we're matching this"

Reply:
I agree with your reasoning. I ended up going with `appStart`. No specific _strong_ reason, but I didn't want `appInit` to be confused with the `workerInitRequest`, since these hooks will be called in also the `functionEnvironmentReloadRequest`.